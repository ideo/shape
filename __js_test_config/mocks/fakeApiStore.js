import fakeUiStore from './fakeUiStore'

const fakeApiStore = ({
  findResult = '',
  findAllResult = [],
  requestResult = { data: {} },
} = {}) => {
  return {
    add: jest.fn().mockName('add'),
    removeAll: jest.fn().mockName('removeAll'),
    find: jest.fn().mockReturnValue(findResult),
    findAll: jest.fn().mockReturnValue(findAllResult),
    request: jest
      .fn()
      .mockReturnValue(Promise.resolve(requestResult))
      .mockName('request'),
    fetch: jest
      .fn()
      .mockReturnValue(Promise.resolve(requestResult))
      .mockName('fetch'),
    fetchRoles: jest
      .fn()
      .mockReturnValue(Promise.resolve())
      .mockName('fetchRoles'),
    sync: jest.fn().mockName('sync'),
    loadCurrentUser: jest.fn().mockReturnValue(Promise.resolve()),
    searchUsersAndGroups: jest
      .fn()
      .mockReturnValue(Promise.resolve({ data: [] })),
    searchUsers: jest
      .fn()
      .mockReturnValue(Promise.resolve({ data: [] })),
    searchCollections: jest.fn().mockReturnValue(Promise.resolve({ data: [] })),
    loadCurrentUserGroups: jest.fn(),
    searchRoles: jest.fn().mockReturnValue(Promise.resolve({ data: [] })),
    setCurrentUserId: jest.fn(),
    findOrganizationById: jest.fn().mockReturnValue({ name: 'abc' }),
    fetchThreads: jest.fn().mockReturnValue(Promise.resolve()),
    fetchNotifications: jest.fn().mockReturnValue(Promise.resolve()),
    findOrBuildCommentThread: jest
      .fn()
      .mockReturnValue(Promise.resolve({ key: 'x' })),
    findThreadForRecord: jest.fn(),
    loadNextThreadPage: jest.fn(),
    createTemplateInstance: jest.fn(),
    moveCards: jest.fn(),
    linkCards: jest.fn(),
    duplicateCards: jest.fn(),
    checkInMyCollection: jest.fn(),
    unreadActivityCount: 0,

    usersThreadPagesToLoad: 1,
    loadingThreads: false,
    hasOlderThreads: false,

    currentUserId: '1',
    currentUser: {
      id: '1',
      current_user_collection_id: 99,
      current_organization: {
        name: 'Acme Inc',
        deactivated: false,
      },
      organizations: [{ name: 'Acme Inc 1', filestack_file_url: 'test.jpg' }],
      name: 'Johnny Appleseed',
      pic_url_square: 'https://s3.amazonaws.com/pic.png',
      groups: [],
      terms_accepted: true,
      switchOrganization: jest.fn(),
      API_hideHelper: jest.fn(),
    },
    // NOTE: important that this matches currentUserOrganization.slug
    currentOrgSlug: 'org-slug',
    currentUserOrganization: {
      id: '1',
      slug: 'org-slug',
      name: 'Acme Inc',
      primary_group: {
        name: 'Acme Inc',
      },
    },
    unreadNotifications: [],
    recentNotifications: new Map(),
    currentThreads: [
      {
        id: '1',
        key: 'abc-1',
        record: { name: 'abc' },
        comments: [],
        users_thread: {
          subscribed: true,
        },
      },
      {
        id: '2',
        key: 'abc-2',
        record: { name: 'xyz' },
        comments: [],
        users_thread: {
          subscribed: true,
        },
      },
    ],
    collections: [],
    items: [],

    uiStore: fakeUiStore,
    undoStore: {},
    routingStore: {},
  }
}

export default fakeApiStore
