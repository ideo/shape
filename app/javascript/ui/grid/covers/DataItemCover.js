import PropTypes from 'prop-types'
import { inject, observer, PropTypes as MobxPropTypes } from 'mobx-react'
import { runInAction, observable } from 'mobx'

import trackError from '~/utils/trackError'
import DataItemCoverCollectionsItems from '~/ui/grid/covers/data-item/DataItemCoverCollectionsItems'
import DataItemCoverDisplayOnly from '~/ui/grid/covers/data-item/DataItemCoverDisplayOnly'

// eslint-disable-next-line react/no-multi-comp
@inject('apiStore')
@observer
class DataItemCover extends React.Component {
  @observable
  targetCollection = null

  componentDidMount() {
    const { collectionFilter } = this.props.item
    if (collectionFilter && collectionFilter.target) {
      this.loadTargetCollection(collectionFilter.target)
    }
  }

  async loadTargetCollection(target) {
    const { apiStore } = this.props
    try {
      const res = await apiStore.fetch('collections', target)
      runInAction(() => {
        this.targetCollection = res.data
      })
    } catch (e) {
      trackError(e)
    }
  }

  render() {
    const { item, card } = this.props
    if (item.isReportTypeCollectionsItems) {
      return <DataItemCoverCollectionsItems item={item} card={card} />
    }
    return <DataItemCoverDisplayOnly item={item} card={card} />
  }
}

DataItemCover.propTypes = {
  item: MobxPropTypes.objectOrObservableObject.isRequired,
  card: MobxPropTypes.objectOrObservableObject.isRequired,
  // eslint-disable-next-line react/no-unused-prop-types
  datasetLength: PropTypes.number,
}

DataItemCover.defaultProps = {
  datasetLength: 1,
}

DataItemCover.wrappedComponent.propTypes = {
  apiStore: MobxPropTypes.objectOrObservableObject.isRequired,
}

export default DataItemCover
