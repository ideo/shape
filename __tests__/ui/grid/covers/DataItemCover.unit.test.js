import DataItemCover from '~/ui/grid/covers/DataItemCover'
import expectTreeToMatchSnapshot from '#/helpers/expectTreeToMatchSnapshot'
import {
  fakeDataItemCollectionsItemsAttrs,
  fakeDataItemRecordAttrs,
} from '#/mocks/data'
import fakeApiStore from '#/mocks/fakeApiStore'

const props = {}
let wrapper, render
const apiStore = fakeApiStore()
describe('DataItemCover', () => {
  beforeEach(() => {
    props.apiStore = apiStore
    props.item = fakeDataItemCollectionsItemsAttrs
    props.card = { id: 1, record: props.item, width: 1, height: 1 }
    render = () =>
      (wrapper = shallow(<DataItemCover.wrappedComponent {...props} />))
    render()
  })

  it('renders snapshot', () => {
    expectTreeToMatchSnapshot(wrapper)
  })

  describe('item with "report_type_collections_and_items"', () => {
    beforeEach(() => {
      props.item = fakeDataItemCollectionsItemsAttrs
      props.card.record = props.item
      wrapper.setProps(props)
      render()
    })

    it('renders DataItemCoverCollectionsItems', () => {
      expect(wrapper.find('DataItemCoverCollectionsItems').exists()).toBe(true)
      expect(wrapper.find('DataItemCoverDisplayOnly').exists()).toBe(false)
    })
  })

  describe('item with "report_type_record"', () => {
    beforeEach(() => {
      props.item = fakeDataItemRecordAttrs
      props.card.record = props.item
      wrapper.setProps(props)
      render()
    })

    it('renders DataItemCoverCollectionsItems', () => {
      expect(wrapper.find('DataItemCoverDisplayOnly').exists()).toBe(true)
      expect(wrapper.find('DataItemCoverCollectionsItems').exists()).toBe(false)
    })
  })
})
