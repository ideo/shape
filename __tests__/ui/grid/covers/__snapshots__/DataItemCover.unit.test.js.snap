// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataItemCover renders snapshot 1`] = `
<DataItemCoverCollectionsItems
  card={
    Object {
      "height": 1,
      "id": 1,
      "record": Object {
        "breadcrumb": Array [
          Array [
            "collections",
            1,
            "Some collection",
          ],
          Array [
            "items",
            1,
            "my item",
          ],
        ],
        "can_edit": false,
        "can_view": true,
        "data_settings": Object {
          "d_measure": "participants",
          "d_timeframe": "month",
        },
        "datasets": Array [
          Object {
            "chart_type": "area",
            "data": Array [
              Object {
                "date": "2018-07-10",
                "value": 10,
              },
              Object {
                "date": "2018-08-10",
                "value": 25,
              },
              Object {
                "date": "2018-09-10",
                "value": 30,
              },
            ],
            "dataWithDates": Array [
              Object {
                "date": 2018-07-10T00:00:00.000Z,
                "value": 10,
              },
              Object {
                "date": 2018-08-10T00:00:00.000Z,
                "value": 25,
              },
              Object {
                "date": 2018-09-10T00:00:00.000Z,
                "value": 30,
              },
            ],
            "description": "A description",
            "identifier": "question",
            "measure": "participants",
            "order": 0,
            "tiers": Array [],
            "timeframe": "month",
          },
        ],
        "fullyLoaded": true,
        "id": "1",
        "inherited_tag_list": Array [],
        "internalType": "items",
        "isReportTypeCollectionsItems": true,
        "isReportTypeNetworkAppMetric": false,
        "isReportTypeRecord": false,
        "measure": Object {
          "name": "Participants",
        },
        "name": "My Cool Item",
        "parentPath": "/",
        "parent_collection_card": Object {
          "API_archive": [MockFunction],
          "API_create": [MockFunction],
          "API_linkToMyCollection": [MockFunction],
          "assign": [MockFunction],
          "assignRef": [MockFunction],
          "beginReplacing": [MockFunction],
          "col": 1,
          "create": [MockFunction],
          "disableMenu": [MockFunction],
          "height": 1,
          "id": "11",
          "image_contain": false,
          "item": Object {},
          "maxHeight": 1,
          "maxWidth": 1,
          "order": 1,
          "patch": [MockFunction],
          "record": Object {},
          "reference": false,
          "reselectOnlyEditableRecords": [MockFunction],
          "reselectOnlyMovableCards": [MockFunction],
          "row": 0,
          "save": [MockFunction],
          "section_type": null,
          "update": [MockFunction],
          "width": 1,
        },
        "primaryDataset": Object {
          "chart_type": "bar",
          "data": Array [],
          "data_items_datasets_id": 100,
          "identifier": "dataset",
          "measure": "participants",
          "name": "dataset",
          "order": 0,
          "question_type": null,
          "selected": true,
          "test_collection_id": null,
          "tiers": Array [],
          "timeframe": "ever",
          "total": 1,
        },
        "quill_data": null,
        "report_type": "report_type_collections_and_items",
        "type": "Item::DataItem",
      },
      "width": 1,
    }
  }
  item={
    Object {
      "breadcrumb": Array [
        Array [
          "collections",
          1,
          "Some collection",
        ],
        Array [
          "items",
          1,
          "my item",
        ],
      ],
      "can_edit": false,
      "can_view": true,
      "data_settings": Object {
        "d_measure": "participants",
        "d_timeframe": "month",
      },
      "datasets": Array [
        Object {
          "chart_type": "area",
          "data": Array [
            Object {
              "date": "2018-07-10",
              "value": 10,
            },
            Object {
              "date": "2018-08-10",
              "value": 25,
            },
            Object {
              "date": "2018-09-10",
              "value": 30,
            },
          ],
          "dataWithDates": Array [
            Object {
              "date": 2018-07-10T00:00:00.000Z,
              "value": 10,
            },
            Object {
              "date": 2018-08-10T00:00:00.000Z,
              "value": 25,
            },
            Object {
              "date": 2018-09-10T00:00:00.000Z,
              "value": 30,
            },
          ],
          "description": "A description",
          "identifier": "question",
          "measure": "participants",
          "order": 0,
          "tiers": Array [],
          "timeframe": "month",
        },
      ],
      "fullyLoaded": true,
      "id": "1",
      "inherited_tag_list": Array [],
      "internalType": "items",
      "isReportTypeCollectionsItems": true,
      "isReportTypeNetworkAppMetric": false,
      "isReportTypeRecord": false,
      "measure": Object {
        "name": "Participants",
      },
      "name": "My Cool Item",
      "parentPath": "/",
      "parent_collection_card": Object {
        "API_archive": [MockFunction],
        "API_create": [MockFunction],
        "API_linkToMyCollection": [MockFunction],
        "assign": [MockFunction],
        "assignRef": [MockFunction],
        "beginReplacing": [MockFunction],
        "col": 1,
        "create": [MockFunction],
        "disableMenu": [MockFunction],
        "height": 1,
        "id": "11",
        "image_contain": false,
        "item": Object {},
        "maxHeight": 1,
        "maxWidth": 1,
        "order": 1,
        "patch": [MockFunction],
        "record": Object {},
        "reference": false,
        "reselectOnlyEditableRecords": [MockFunction],
        "reselectOnlyMovableCards": [MockFunction],
        "row": 0,
        "save": [MockFunction],
        "section_type": null,
        "update": [MockFunction],
        "width": 1,
      },
      "primaryDataset": Object {
        "chart_type": "bar",
        "data": Array [],
        "data_items_datasets_id": 100,
        "identifier": "dataset",
        "measure": "participants",
        "name": "dataset",
        "order": 0,
        "question_type": null,
        "selected": true,
        "test_collection_id": null,
        "tiers": Array [],
        "timeframe": "ever",
        "total": 1,
      },
      "quill_data": null,
      "report_type": "report_type_collections_and_items",
      "type": "Item::DataItem",
    }
  }
/>
`;
